pkg/kubecli/vmi.go

```diff
@@ -258,24 +258,21 @@ func (v *vmis) SerialConsole(name string, timeout time.Duration) (StreamInterfac
 	isWaiting := true
 
 	go func() {
-		con, err := v.asyncSubresourceHelper(name, "console")
-		for err != nil && isWaiting {
-			if asyncSubresourceError, ok := err.(*AsyncSubresourceError); ok {
-				if asyncSubresourceError.GetStatusCode() == http.StatusBadRequest {
-					// Sleep to prevent denial of service on the api server
-					time.Sleep(1 * time.Second)
-					con, err = v.asyncSubresourceHelper(name, "console")
-				} else {
-					connectionChan <- connectionStruct{con: nil, err: asyncSubresourceError}
+		for {
+			con, err := v.asyncSubresourceHelper(name, "console")
+			if err != nil {
+				asyncSubresourceError, ok := err.(*AsyncSubresourceError)
+				if !isWaiting || !ok || asyncSubresourceError.GetStatusCode() != http.StatusBadRequest {
```

Comment:
can we simplify this logic. having trouble following what the intent is here. 

Also, we're using a boolean as a flag to communicate between two goroutines with the 'isWaiting' boolean. I'm pretty sure that isn't safe. From what I remember, there's no guarantee for when the goroutine reading the isWaiting boolean will get the update when it is set to 'false' in the other goroutine.

Reply:
to be honest, I found it more complicated to understand recursive logic(previous implementation), I can add the comment on the `if` block to explain which conditions do not suit, will it be enough?
And you are right about `isWaiting` boolean is not atomic variable, just copied it from the previous implementation but will fix it.

Reply:
can we just put the timeout ticker and select statement in the goroutine and get rid of the isWaiting boolean entirely?